2단계 - 지하철 노선 기능 과제 PR 요청 드립니다!

taewonh opened the pull request:
안녕하세요!
2단계 지하철 노선 기능 과제가 완료되어 PR 요청 드립니다.
이번 과제에서 중점적으로 작업한 내용들은 아래와 같습니다!

- 기능 요구사항 반영 및 각 기능별 인수 테스트 코드 작성
  - 지하철 노선 생성
  - 지하철 노선 목록 조회
  - 지하철 노선 조회
  - 지하철 노선 수정
  - 지하철 노선 삭제
- 지하철 노선 요구사항 추가에 따른 Entity 설계 및 적용
  - 해당 부분은 이번 요구사항에 존재하진 않았지만 작업을 진행하다 보니 Entity 설계까지 동시에 진행되어 버렸네요..
  - Line, Section Entity 추가
 
작업하다보니 DTO -> Entity 변환과 DTO와 Entity를 사용하는 영역 (컨트롤러, 서비스) 을 결정하는 것, 그리고 각 레이어의 역할에 대해 계속 고민이 되더라구요.. 그러다 보니 코드가 마음에 안들어서 수정도 여러번 하게 되었구요..
관련한 내용으로 고민한 블로그들도 꽤 많이 보여서 개발자의 취향 차이라는 생각도 들어서 제 나름대로 규칙을 정하고 진행해봤습니다.
- DTO
  - Data Transaction Object 라고 하는 명칭의 의미를 지키기 위해 Controller에서 받아서 Service로 전달하는데에만 사용. 
  - Service 영역에서는 전달받은 DTO를 이용해서 Entity 화 진행
  - Controller 영역에서는 Entity에 대한 사용 자체 금지. 알아서도 안됨
- 동일 계층간 통신 불가능 (ex. service 간 호출, controller 간 호출)
  - layer간 호출을 controller -> service -> repository 순 으로만 허용

혹시 이러한 부분 관련해서 평소에 리뷰어님께서 생각하시거나 지키시는 철학같은 것이 있는지 궁금합니다!

taewonh committed:
refactor(StationacceptanceTest): API 호출 메소드 명칭 변경

- Step1 피드백 반영

taewonh committed:
docs(README.md): 기능 요구사항 작성

taewonh committed:
test(StationAcceptanceTest): 추가할 기능에 대한 인수 테스트 샘플 코드 작성

taewonh committed:
feat: Line, Section Entity 및 Repository 추가

- Line (노선)
  - 노선명, 색상, 구간 목록을 멤버로 가짐
- Section (구간)
  - 라인, 상행역, 하행역, 역간 거리를 멤버로 가짐

taewonh committed:
feat: 지하철 노선 생성 기능 추가

- API 호출 및 응답에 사용하는 LineRequest, LineResponse 추가
- LineRequest를 이용하여 Line 및 Section을 생성하여 등록 처리

taewonh committed:
feat: 지하철 노선 목록 조회 기능 추가

- 추가로 기존 StationAcceptanceTest 에 작성하던 노선 관련 테스트 코드를 별도 클래스(LineAcceptanceTest)로 분리

taewonh committed:
refactor: 전체 코드 리팩토링

- Line 인수 테스트 코드 고도화
  - 라인 생성시 테스트 데이터 비교를 용이하게 하기 위해 Station 객체를 파라미터로 사용하는 방식으로 변경
  - 기존 테스트 코드의 print 제거 및 assert 추가
  - 코드 리팩토링
- SectionService 제거, 이후 추가 필요시 재 추가 예정
- LineRequest 명칭 변경, 앞으로 API 별로 세분화된 Dto를 추가할 예정으로 호출할 API의 의미가 담겨있는 Dto 명칭으로 변경
- Dto 영역을 컨트롤러에서 사용하지 않고 서비스 영역에서도 사용하도록 변경
  - 사용자 I/O를 담당하는 컨트롤러에서 도메인에 직접 접근하는 것을 방지하기 위함

taewonh committed:
feat: 지하철 노선 조회 기능 추가

taewonh committed:
feat: 지하철 노선 수정 기능 추가

taewonh committed:
feat: 지하철 노선 삭제 기능 추가

taewonh committed:
test: 인수 테스트 코드 격리 적용

- DatabaseCleanup 서비스 추가 후 모든 테이블을 truncate 하는 기능을 매 테스트 코드 실행시 호출하는 방식 적용

dhmin5693 commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -0,0 +1,66 @@
+package nextstep.subway.domain;
```
http://www.javapractices.com/topic/TopicAction.do?Id=205
패키지 구조를 도메인 단위로 변경해보시기 바랍니다.
코드의 응집성 부분에 장점이 있습니다.

이번 미션은 도메인을 나누기가 꽤 명확한 편입니다.
지금 연습해보면 좋은 타이밍이라고 생각되어 의견드립니다.

taewonh replied:
한번 적용해보고 싶은 구조네요! 그런데 도메인 단위로 전체 패키지 구조가 변경했을 때 코드 확인하시는데 문제는 없으실까요..? 지난번 과제때 패키지 명칭을 한번에 변경했다가 코드 확인하기 어렵다는 피드백을 한번 받은적이 있어서요! 만약 문제가 없다고 하시면 변경해보도록 하겠습니다!

dhmin5693 replied:
괜찮습니다. 저는 ATDD 미션에서 항상 이렇게 피드백을 드리고 있습니다.

dhmin5693 commented on src/main/java/nextstep/subway/application/DatabaseCleanup.java:
```diff
@@ -0,0 +1,49 @@
+package nextstep.subway.application;
```
java 아래가 아니라 test 아래에 있어야 맞지 않을까요?
혹시라도 잘못 실행되면 정말정말 큰일나는 코드입니다 😱

테스트 용으로 사용할 목적이기도 하지만 테스트로 보내면 프로덕션 코드에서 아예 참조가 안됩니다.

taewonh replied:
앗... 실수가 있었네요... 해당 부분은 바로 변경 처리하도록 하겠습니다..!

dhmin5693 commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,200 @@
+    @LocalServerPort
+    int port;
+
+    @BeforeEach
+    public void setUp() {
+        if (RestAssured.port == RestAssured.UNDEFINED_PORT) {
+            RestAssured.port = port;
+        }
+        databaseCleanup.execute();
+    }
```
이 설정은 노선 인수 테스트와 정거장 인수 테스트에서 똑같이 쓰이고 있어요.

`AcceptanceTest` 라는 별도의 설정 클래스로 빼내고 인수테스트에서 `AcceptanceTest` 클래스를 상속받는 형식으로 만들어보세요.
설정 중복을 줄일 수 있습니다.

taewonh replied:
넵넵 말씀하신대로 테스트 관련 설정은 상위 클래스로 추출하고 하위 테스트 클래스에서는 테스트 메소드만 정의해서 사용하는 방식으로 변경 처리하겠습니다!

dhmin5693 commented:
구조나 설계 상의 규칙에 대해서 고민하는 일은 되게 건설적이고 좋은 일입니다.
그 결과 합리적인 결론을 도출하신 걸로 보이네요.

제 견해도 조금 밝혀보겠습니다.

---

## DTO

DTO는 완전 외부에서 들어오는 데이터를 정의하는 역할만 주기엔 그 이름의 범위가 넓습니다.
어느 레이어에 있던 정하기에 따라 데이터를 어디까지 전달해줄지를 정해줄 수도 있어요.

예를 들자면 컨트롤러로 들어오는 범위와 컨트롤러에서 서비스로 진입하는 범위를 따로 설정해줄 수도 있는거죠.
조금 과해보일 수 있지만 DTO 등을 더 좁은 범위에서 사용하면 코드의 결합성을 줄일 수 있는 장점도 있습니다.
육각형 아키텍처를 사용한다면 레이어가 다른 경우 무조건 DTO로 넘겨주기도 해요.
(육각형 아키텍처는 시간이 괜찮으실 때 따로 공부해보세요)

## Entity

엔티티 사용 범위도 잘 정의해주셨습니다.
여기도 마찬가지로 사용하기에 따라 서비스에서도 제한하는 방법도 있습니다.
조금 더 정확한 근거는 `레이어드 아키텍처` 의 구조를 공부해보시면 좋습니다.

### 레이어드 아키텍처

레이어드 아키텍처에서는 크게 표현 영역, 서비스 영역, (도메인 영역), 영속 영역으로 분류됩니다.
스프링을 사용해보셨다면 자연스럽게 사용하는 구조이기도 해요.
이 구조에서 지켜주어야 하는 가장 큰 특징은 하위 영역은 보호되어야 한다는 겁니다. 
이 규칙은 하위 영역에서 상위 영역을 참조하지 않는 방법으로 쉽게 지킬 수 있습니다.
`import` 에 포함만 되지 않으면 돼요.

또한 각 영역의 생명 주기도 나름대로 정해볼 수 있습니다.
인접한 두 영역은 상위가 하위를 참조해야만 접근이 가능하니 논할 대상이 아닙니다.

도메인 영역은 표현 영역에서도 사용할 수 있지만 정하기에 따라서 서비스까지만 존재하도록 할 수도 있습니다.
영속 영역도 마찬가지가 될 수 있겠죠?

### Entity의 범위

엔티티는 도메인과 영속 영역에 애매하게 걸쳐있습니다.
도메인의 특징도 가지고 있지만, 엔티티의 가장 큰 특징은 테이블을 자바 코드로 정의했다는 것이니까요.
굳이 따지자면 영속 영역의 역할이 조금 더 크다고는 생각합니다.
그런 관점에선 디비에 접근이 가능한 repository는 영속 영역의 코드를 별도로 만들고, 그 코드 내부에서만 repository를 참조하게 하는 방법도 있습니다.
그러면 엔티티가 영속 영역 밖으로 빠져나오지 않겠죠.

다만 이 부분은 이번 미션에서 강제하지는 않습니다.
이번 미션의 지향점도 아니고 조직마다 정하기 나름이니까요.
스스로 정하신 규칙에 의거하여 코드를 작성해주시면 됩니다.

## 동일 계층 호출 불가능

솔직히 동일 계층 호출하면 되게 편하긴 합니다만, 간혹 의존성이 꼬이거나 사이클이 생기기도 합니다.
그래서 저희 팀에서도 호출은 최대한 덜 하자는 방향으로 코드를 작성합니다.

여기는 적당한 융퉁성을 주어도 좋다고 생각해요.
taewonh님이 적정선을 잘 그어보시면 됩니다.

dhmin5693 commented on src/main/java/nextstep/subway/ui/LineController.java:
```diff
@@ -0,0 +1,52 @@
+
+    @PostMapping("/lines")
+    public ResponseEntity<LineResponse> createLine(@RequestBody CreateLineDto dto) {
+        Line registeredLine = lineService.register(dto);
```
Controller 에서 `Line` Entity를 알고 있습니다.
정하신 규칙에 따르자면 이 부분은 수정하시는게 좋겠네요.

taewonh replied:
해당 부분 역시 말씀하신대로 반영하겠습니다!

dhmin5693 commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Transactional
+    public Line register(CreateLineDto dto) {
```
여기서 `Line`을 반환하게 된다면 컨트롤러로 엔티티가 나가게 될거에요.

taewonh replied:
LineController 부분에서 주신 피드백과 동일하게 반영 처리하겠습니다!

dhmin5693 commented on src/main/java/nextstep/subway/dto/DtoConverter.java:
```diff
@@ -0,0 +1,14 @@
+
+import nextstep.subway.domain.Line;
+
+public class DtoConverter {
```
이 converter는 너무 많은 곳에 의존될 수 있고 다른 클래스를 너무 많이 의존하게 될 가능성이 높아보입니다.
이 클래스 내부의 팩토리 메소드를 각 dto로 옮긴다면 converter와의 결합성은 줄이고 dto의 코드 응집성은 높일 수 있지 않을까요?

dhmin5693 commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -0,0 +1,45 @@
+
+    private final String color;
+
+    private final List<Map<String, Object>> stations;
```
`StationsResponse` 같은 걸로 묶을 수 있지 않을까요?
콜렉션을 중첩해서 사용하면 디버깅이 굉장히 어려우니 지양하시는걸 추천드립니다.

taewonh replied:
넵넵 해당 부분은 말씀하신 대로 적용처리 하였습니다. 다만 수정하다 보니 고민되는 부분이 있는게 외부에 이미 StationResponse(id, name, createdDate, modifiedDate) 가 있어 LineResponse 내부에서 전용으로 사용하는 StationResponse를 private한 내부 클래스로 선언해서 사용하는 방식으로 변경 처리 했습니다. 이 부분도 정답은 없을 것 같긴한데 혹시 조금 더 괜찮은 방법이 있을까요??

dhmin5693 replied:
고민의 이유를 생각해보니 이번 미션에서 API 명세가 빠져있군요. ~~지난 번 기수까지는 미션 내용에 포함이 되어 있던 것으로 기억합니다.~~

~~API에서 보통 호환이 안된다라고 하면 response의 데이터가 모자란 경우가 대부분이에요.~~
~~원래 있던 `StationResponse` 는 이번에 생성한 정거장 정보의 데이터를 모두 가지고 있습니다.~~
~~생성/변경 시간 정도를 더 주고 있네요.~~

~~목적에 알맞는 response를 정확하게 사용하고 분리하는 일은 굉장히 중요합니다.~~
~~다만 이번 미션의 도메인을 생각해보면 정거장은 노선에 속해있으며 정거장 API와 노선 API로부터 전달받는 정거장의 정보는 서로 다를 이유가 없다고 생각해요.~~

~~그래서 이미 있는 `StationResponse` 를 그냥 활용해도 괜찮다고 생각합니다.~~
윗 부분 정정합니다. API 명세가 포함되어 있는데 접혀 있어서 제가 없다고 생각을 해버리고 확인을 못했군요.
`StationResponse` 에 생성/변경 시간이 없고 스펙이 완전 같아서 재사용 해도 되겠네요.

그리고 내가 작업한게 혹시 잘못되었을까? 를 테스트해보시려면 같이 제공된 프론트엔드를 실행해서 기능을 검증해보세요.
그 기능이 제대로 수행되면 문제가 없는 것으로 생각하시면 됩니다.

dhmin5693 commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,200 @@
+        databaseCleanup.execute();
+    }
+
+    private Station saveStation(String name) {
+        return stationRepository.save(new Station(name));
+    }
```
인수 테스트에서 데이터 생성이 필요할 땐 repository에 직접 꽂아 넣으시기보다는 API를 활용해주세요.

taewonh replied:
인수테스트에서 repository 사용은 지양하도록 하겠습니다!

dhmin5693 commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -99,24 +106,24 @@ void getStations() {
     }
 
-    private ValidatableResponse requestRegister(String name) {
+    private ValidatableResponse createStation(String name) {
```
`LineAcceptanceTest` 에서 repository를 사용하는 대신 이 메소드를 사용하는 건 어떤가요?

또한 미션을 진행하다 보면 `createStation` 은 다른 곳에서 많이 쓰이게 됩니다.
`StationTestFixture` 같은 클래스를 생성하고 station 리소스 CRUD 하는 부분을 몰아두면 가독성이 굉장히 좋아집니다.

taewonh replied:
각 도메인 (Station, Line) 에 대한 CRUD는 각 Fixture 클래스 (StationTestFixture, LineTestFixture)를 생성해서 인수 테스트 클래스에서 호출만 하는 방식으로 변경 처리하였습니다! 또한 해당 클래스들은 클래스명칭에 해당하는 도메인에만 CRUD 기능을 제공하기 때문에 메소드에 들어가는 도메인 이름 역시 제거 하였습니다 (ex. createStation -> create)

dhmin5693 commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.stream.Collectors;
+
+@Service
+public class LineService {
```
서비스에 대한 테스트도 작성해주세요.

dhmin5693 commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -0,0 +1,66 @@
+        return color;
+    }
+
+    public List<Map<String, Object>> getStations() {
+        List<Map<String, Object>> stations = sections.stream()
+                .map(section -> section.getUpStation().toMapForOpen())
+                .collect(Collectors.toList());
+        Section lastSection = sections.get(sections.size() - 1);
+        stations.add(lastSection.getDownStation().toMapForOpen());
+        return stations;
+    }
```
도메인 부분 역시 테스트를 작성해주세요.
TDD 미션만 단위 테스트를 작성하는건 아닙니다. (애초에 테스트를 작성한다고 TDD라 볼 수 없습니다)

public으로 외부에 공개한 메소드는 언제, 어디까지 쓰일지 전혀 예상할 수 없는 범위에 놓이게 돼요.
단위 테스트로 동작을 보증해주세요.

특히 조금이라도 복잡해진 public 메소드는 꼭 테스트를 작성해주시기 바랍니다.

dhmin5693 commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,46 @@
+    private Station downStation;
+
+    @Column
+    private Long distance;
```
`Distance` 도 역할을 별도로 부여할 수 있겠네요.

taewonh replied:
넵 현재 요구사항까지는 distance가 특정한 기능은 제공하지 않아 Long 타입의 distance 변수만 가지는 래퍼 클래스만 생성해서 우선 적용 처리했습니다!

dhmin5693 commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -0,0 +1,66 @@
+    @Column(unique = true)
+    private String color;
+
+    @OneToMany(mappedBy = "line", cascade = { CascadeType.PERSIST, CascadeType.REMOVE })
+    private final List<Section> sections = new ArrayList<>();
```
여기는 일급 콜렉션을 활용해보시는건 어떨까요?

dhmin5693 reviewed (changes requested):
taewonh님 안녕하세요.

이번 미션을 위해 정말 많은 고민을 해 주신 것 같습니다.
제 생각을 답변으로 정리해두었으니 참고해보시기 바랍니다.

그리고 미션도 정말 열심히 수행해주신게 눈에 보이네요.
의견을 달아두었으니 한번 확인해보시고 수정 후 다시 요청주세요.

어려운 점 있으시다면 언제든지 DM 부탁드리겠습니다.

taewonh commented:
> 구조나 설계 상의 규칙에 대해서 고민하는 일은 되게 건설적이고 좋은 일입니다. 그 결과 합리적인 결론을 도출하신 걸로 보이네요.
> 
> 제 견해도 조금 밝혀보겠습니다.
> 
> ## DTO
> DTO는 완전 외부에서 들어오는 데이터를 정의하는 역할만 주기엔 그 이름의 범위가 넓습니다. 어느 레이어에 있던 정하기에 따라 데이터를 어디까지 전달해줄지를 정해줄 수도 있어요.
> 
> 예를 들자면 컨트롤러로 들어오는 범위와 컨트롤러에서 서비스로 진입하는 범위를 따로 설정해줄 수도 있는거죠. 조금 과해보일 수 있지만 DTO 등을 더 좁은 범위에서 사용하면 코드의 결합성을 줄일 수 있는 장점도 있습니다. 육각형 아키텍처를 사용한다면 레이어가 다른 경우 무조건 DTO로 넘겨주기도 해요. (육각형 아키텍처는 시간이 괜찮으실 때 따로 공부해보세요)
> 
> ## Entity
> 엔티티 사용 범위도 잘 정의해주셨습니다. 여기도 마찬가지로 사용하기에 따라 서비스에서도 제한하는 방법도 있습니다. 조금 더 정확한 근거는 `레이어드 아키텍처` 의 구조를 공부해보시면 좋습니다.
> 
> ### 레이어드 아키텍처
> 레이어드 아키텍처에서는 크게 표현 영역, 서비스 영역, (도메인 영역), 영속 영역으로 분류됩니다. 스프링을 사용해보셨다면 자연스럽게 사용하는 구조이기도 해요. 이 구조에서 지켜주어야 하는 가장 큰 특징은 하위 영역은 보호되어야 한다는 겁니다. 이 규칙은 하위 영역에서 상위 영역을 참조하지 않는 방법으로 쉽게 지킬 수 있습니다. `import` 에 포함만 되지 않으면 돼요.
> 
> 또한 각 영역의 생명 주기도 나름대로 정해볼 수 있습니다. 인접한 두 영역은 상위가 하위를 참조해야만 접근이 가능하니 논할 대상이 아닙니다.
> 
> 도메인 영역은 표현 영역에서도 사용할 수 있지만 정하기에 따라서 서비스까지만 존재하도록 할 수도 있습니다. 영속 영역도 마찬가지가 될 수 있겠죠?
> 
> ### Entity의 범위
> 엔티티는 도메인과 영속 영역에 애매하게 걸쳐있습니다. 도메인의 특징도 가지고 있지만, 엔티티의 가장 큰 특징은 테이블을 자바 코드로 정의했다는 것이니까요. 굳이 따지자면 영속 영역의 역할이 조금 더 크다고는 생각합니다. 그런 관점에선 디비에 접근이 가능한 repository는 영속 영역의 코드를 별도로 만들고, 그 코드 내부에서만 repository를 참조하게 하는 방법도 있습니다. 그러면 엔티티가 영속 영역 밖으로 빠져나오지 않겠죠.
> 
> 다만 이 부분은 이번 미션에서 강제하지는 않습니다. 이번 미션의 지향점도 아니고 조직마다 정하기 나름이니까요. 스스로 정하신 규칙에 의거하여 코드를 작성해주시면 됩니다.
> 
> ## 동일 계층 호출 불가능
> 솔직히 동일 계층 호출하면 되게 편하긴 합니다만, 간혹 의존성이 꼬이거나 사이클이 생기기도 합니다. 그래서 저희 팀에서도 호출은 최대한 덜 하자는 방향으로 코드를 작성합니다.
> 
> 여기는 적당한 융퉁성을 주어도 좋다고 생각해요. taewonh님이 적정선을 잘 그어보시면 됩니다.

taewonh committed:
refactor(DatabaseCleanup): 실행 영역 변경

- Production 영역에서는 사용될 필요가 없는 서비스 코드로써 test 시에만 사용할 수 있도록 변경 처리

taewonh committed:
test: 인수 테스트 내부의 설정 관련 항목을 상위 클래스로 추출

- 실제 테스트 코드를 작성하는 하위 테스트 클래스는 테스트만을 담당하는 방식으로 리팩토링

taewonh committed:
refactor: Line register return 타입 변경

- Line이 아닌 LineResponse로 타입 변경
- Service로 부터 return되는 항목을 도메인 객체가 아닌 DTO로 변경함으로써, Controller 영역에서 도메인에 접근하는 것을 방지하기 위해 적용

taewonh committed:
refactor: Section List 일급 컬렉션 적용

taewonh committed:
refactor: LineResponse의 중첩 콜렉션 제거

- Station 정보(id, name)을 담고 있는 Map을 클래스로 추출
- 추출한 StationResponse는 해당 LineResponse 내부에서만 사용 가능한 inner 클래스로 적용
- inner 클래스 리스트를 생성하는 내부 메소드 역시 외부에서 사용 불가능하도록 private 처리

taewonh committed:
test: 각 인수테스트에서 사용하는 crud 로직을 별도의 Fixture 클래스로 분리

- 각 Fixture 클래스는 static 메소드로 crud 기능 제공

taewonh committed:
refactor: Section의 distance를 별도의 래퍼 클래스로 구성

taewonh committed:
refactor: DtoConverter 제거

- 제거되는 DtoConverter에서 제공하는 toEntity 기능은 각 Dto 내부에서 역할하도록 변경

taewonh committed:
test: 서비스, 도메인 테스트 추가

- 노선, 지하철역 서비스 테스트 추가
- 구간 목록, 지하철 서비스 테스트 추가

dhmin5693 commented on src/test/java/nextstep/subway/application/DatabaseCleanup.java:
```diff
@@ -0,0 +1,49 @@
+package nextstep.subway.application;
```
잘 옮겨주셨네요 :)
application package에 속할 내용은 아닐 수도 있습니다.
이 패키지 안에 들어 있는 다른 테스트와는 조금 상이한 성격을 가지고 있어요.

taewonh replied:
음 일단 util 패키지로 변경처리 했는데 비슷한 느낌일지는 확실히는 모르겟군요... ㅜ

taewonh replied:
해당 부분뿐만 아니라 패키지 전체 구조를 변경하면서 모든 클래스들에 대해서 기본적으로 package-private 방식으로 고수하도록 하고, 그 외 다 패키지 클래스로의 접근이 필요한 경우에 대해서만 public 적용처리 까지 완료되었습니다!

dhmin5693 replied:
넵, 적당한 이름이면 괜찮습니다 :)
그 외 세심한 부분까지 신경써주셨군요. 👍

dhmin5693 commented on src/test/java/nextstep/subway/fixture/LineTestFixture.java:
```diff
@@ -0,0 +1,60 @@
+package nextstep.subway.fixture;
```
`TextFixture` 분리 👍 👍 

테스트도 도메인 단위로 패키지를 나누면 더 좋겠네요!

dhmin5693 commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -1,33 +1,18 @@
-        }
-    }
+@DisplayName("지하철역 인수 테스트")
+public class StationAcceptanceTest extends AcceptanceTest {
```
👍

dhmin5693 commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,122 @@
+    @Test
+    void getLines() {
+        // Given
+        Station 강남역 = StationTestFixture.create("강남역").extract().as(Station.class);
+        Station 양재역 = StationTestFixture.create("양재역").extract().as(Station.class);
```
이 부분은 API로 받아온 response의 정보입니다.

테스트 입장으로 보자면 API를 호출했으니 `StationResponse` 의 스펙은 알고 있어야 맞겠지만
`Station` entity의 정보는 알 수가 없으며 알아서도 안됩니다.

`Station` 이 아니라 `StationResponse`를 활용하도록 변경해주세요.
`Line`도 같은 문제가 있다면 같이 수정해주세요.

dhmin5693 reviewed (changes requested):
피드백 잘 반영해주셨네요 :)

다른 부분 점검이 완료되어 구조 변경으로 인해 코드 변경량이 많아보여도 괜찮으니 적용해주시고 다시 요청 부탁드리겠습니다.
몇 가지 더 남겼으니 이 부분도 확인해주세요!

taewonh committed:
refactor: LineResponse 시 StationResponse 재사용 적용

- LineResponse 내부에 선언된 StationResponse 제거 및 기존 사용되고 있던 StationResponse를 사용하도록 변경

taewonh committed:
test: acceptance 테스트시 도메인 객체를 사용하지 않고 Response 전용 DTO를 사용하도록 변경

taewonh committed:
refactor: 패키지 구성 방식 변경

- 계층형 아키텍처로 구성된 패키지를 기능 아키텍처 방식으로 패키지 구조 변경
- 모든 클래스들은 기본적으로 package-private 방식을 고수하고 타 파일에서 접근을 필요로하는 경우에만 public으로 변경

dhmin5693 reviewed (approved):
피드백 반영 너무 잘해주셨어요. 💯 
3단계 진행해주세요.
수고 많으셨습니다!
